### PR TITLE
[MMT] - New glossary endpoints

### PR DESCRIPTION
**Pull request by mauretto78**
ASANA: https://app.asana.com/0/1134617950425092/1205740284473357

| Documentation content |
| --- |
| 1. [New endpoints](#head1) |
| 2. [Project creation](#head2) |
| 3. [Retrieving mmt_glossaries value](#head3) |

## <a name="head1"></a>1. NEW ENDPOINTS

Seven new endpoints were added:

| ROUTE                                             | VERB | DATA | DESCRIPTION                  |
|---------------------------------------------------|------|------|------------------------------|
| /api/v3/mmt/[:engineId]/keys                      | GET  | x-www-form-urlencoded  | Fetch the list of memories |
| /api/v3/mmt/[:engineId]/create-memory-and-import-glossary           | POST | form-data | Upload a glossary            |
| /api/v3/mmt/[:engineId]/import-glossary           | POST | form-data | Upload a glossary            |
| /api/v3/mmt/[:engineId]/modify-glossary           | POST | raw(JSON) |Update an existing glossary  |
| /api/v3/mmt/[:engineId]/create-memory | POST | x-www-form-urlencoded | Create a new memory          |
| /api/v3/mmt/[:engineId]/update-memory/[:memoryId] | POST | x-www-form-urlencoded  | Update a memory         |
| /api/v3/mmt/[:engineId]/delete-memory/[:memoryId] | GET | x-www-form-urlencoded  | Delete a memory             |
| /api/v3/mmt/[:engineId]/job-status/[:uuid]                      | GET  | x-www-form-urlencoded  | Fetch the import job status |

**IMPORTANT NOTE** For every endpoint, the `engineId` is required. An error is thrown:
- if this does not correspond to an existing engine 
- if the engine is not a valid instance of MMT 
- if the engine does not belong to the logged user

### /api/v3/mmt/[:engineId]/keys

This endpoint returns a list of memories. 

Example:

```json
[
    {
        "id": 197983,
        "name": "TM esame",
        "has_glossary": false
    },
    {
        "id": 197996,
        "name": "TM corso",
        "has_glossary": false
    },
    {
        "id": 198430,
        "name": "MyMemory eb90b0a88870372f2599",
        "has_glossary": false
    }
]
```

Is possible to filter out the results with two parameters:

- `has_glossary` - allowed values 1 or 0
- `q` - filter by name

Example of call:

```/api/v3/mmt/[:engineId]/keys?q=prova&has_glossary=1```

to filter only memories with glossary and name like 'prova'.

### /api/v3/mmt/[:engineId]/create-memory-and-import-glossary

Use this import to create a new memory and upload a glossary.

Required params:

- `glossary` - the glossary file. The file must be an xlsx file. 
- `name` - the new memory name 

If no errors occur, this response is returned:

```json
{
    "id": "00000000-0000-0000-0000-0000003ec4dc",
    "memory": 324214,
    "size": 6,
    "begin": 962155779,
    "end": 962155784,
    "dataChannel": 0,
    "progress": 0
}
```

### /api/v3/mmt/[:engineId]/import-glossary

Use this endpoint to upload a glossary on an existing memory.

Required params:

- `glossary` - the glossary file. The file must be an xlsx file. 
- `memoryId` - the memory Id 

If no errors occur, this response is returned:

```json
{
    "id": "00000000-0000-0000-0000-0000003ec4d5",
    "memory": 321140,
    "size": 6,
    "begin": 962155772,
    "end": 962155777,
    "dataChannel": 0,
    "progress": 0
}
```

The id can be used to check the import job status.

### /api/v3/mmt/[:engineId]/modify-glossary

Use this endpoint to modify a glossary.

The request MUST be JSON. 

Example:

```json
{
    "memory": 324214,
    "size": 1,
    "begin": 97594382,
    "end": 97594382,
    "dataChannel": 1,
    "progress": 0,
    "id": "80000001-0000-0000-0000-000005d12c0e"
}
```

If you are using `equivalent` type, `tuid` param MUST be used. Example:

```json
{
    "tuid": 12345,
    "memoryId": 96447,
    "type": "equivalent",
    "terms": [
        {
            "term": "test",
            "language": "en"
        },
        {
            "term": "prova",
            "language": "it"
        }
    ]
}
```

If no errors occur, this response is returned:

```json
{
    "memory": 323048,
    "size": 1,
    "begin": 97095013,
    "end": 97095013,
    "dataChannel": 1,
    "progress": 0,
    "id": "80000001-0000-0000-0000-000005c98d65"
}
```

The id can be used to check the import job status.

### /api/v3/mmt/[:engineId]/create-memory

Use this endpoint to create a new memory.

Required params:

- `name` - the memory name
- `description` (optional) - the description of the memory
- `external_id` (optional) - the memory key from [MyMemory](https://mymemory.translated.net/), prefixed with x_mm-. Its content will be automatically imported

If no errors occur, this response is returned:

```json
{
    "id": 324235,
    "owner": {
        "user": 2410,
        "client": 0
    },
    "key": "mk_rfN1+GgNlzYL2OU-0LkC4Q",
    "creationDate": "2023-11-14T15:25:31+00:00",
    "name": "dasdsa",
    "hasGlossary": false
}
```

### /api/v3/mmt/[:engineId]/update-memory/[:memoryId]

Use this endpoint to update a memory name.

Required params:

- `name` - the memory name to be updated

If no errors occur, this response is returned:

```json
{
    "id": 324235,
    "owner": {
        "user": 2410,
        "client": 0
    },
    "key": "mk_rfN1+GgNlzYL2OU-0LkC4Q",
    "creationDate": "2023-11-14T15:25:31+00:00",
    "name": "dasdsadddd",
    "hasGlossary": 0
}
```

### /api/v3/mmt/[:engineId]/delete-memory/[:memoryId]

Use this endpoint to delete a memory.

If no errors occur, this response is returned:

```json
{
    "id": 324235,
    "owner": {
        "user": 2410,
        "client": 0
    },
    "key": "mk_rfN1+GgNlzYL2OU-0LkC4Q",
    "creationDate": "2023-11-14T15:25:31+00:00",
    "name": "dasdsadddd",
    "hasGlossary": 0
}
```

### /api/v3/mmt/[:engineId]/job-status/[:uuid]

Use this endpoint to fetch the import job status.

If no errors occur, this response is returned:

```json
{
    "id": "00000000-0000-0000-0000-0000003e9dc3",
    "memory": 321140,
    "size": 6,
    "begin": 960994456,
    "end": 960994461,
    "dataChannel": 0,
    "progress": 1
}
```

`progress` set to 1 means that the job was successfully completed.

### Error handling

In case of error, MMT will return an 4xx or 5xx http code. 

The error message has this form:

```json
{
  "status": 400,
  "error": {
    "type": "UnsupportedLanguageException",
    "message": "Language pair not supported: en > xd"
  }
}
```

For more info read official docs: https://www.modernmt.com/api/?shell#management

## <a name="head2"></a>2. PROJECT CREATION

The project creation now supports `mmt_glossaries` parameter.

This param is subjected to validation:

- the param must be a valid JSON string
- the form of JSON string must respect like structure: `{"ignore_glossary_case": false, "glossaries": [1,2,3,4]}`

## <a name="head3"></a>3. RETRIEVING `mmt_glossaries` VALUE

Use this standard API to get `mmt_glossaries` value:

```/api/v3/jobs/23/64ab53e57c0a/metadata```

Example:

```json
{
  "project": {
    "features": "mmt,translation_versions,review_extended,second_pass_review",
    "mmt_glossaries": { 
        "glossaries": [1, 2, 3, 4], 
        "ignore_glossary_case": true 
      }
  },
  "job": {},
  "files": [
    {
      "id": "23",
      "filename": "modificato.xlf",
      "data": {}
    }
  ]
}
```

### CODE REVIEW

File: lib/Controller/createProjectController.php [hunk 1]
```diff
@@ -128,6 +133,28 @@ public function __construct() {
...
+        if($this->mt_engine > 0 and $this->isLoggedIn()){
+            try {
+                EngineValidator::engineBelongsToUser($this->mt_engine, $this->user->uid);
+            } catch (Exception $exception){
+                $this->result[ 'errors' ][] = [ "code" => -6, "message" => $exception->getMessage() ];
+            }
+        }
+
+        // validate mmt_glossaries string
+        if ( !empty( $this->postInput[ 'mmt_glossaries' ] ) and $this->isLoggedIn() ) {
```
Comment (Ostico) [inline lib/Controller/createProjectController.php:146]: the check for user login is redundant, can be removed

File: lib/Utils/Engines/MMT.php [hunk 2]
```diff
@@ -51,14 +51,6 @@ public function __construct( $engineRecord ) {
 
     }
 
-    /**
-     * @return EnginesModel_EngineStruct
-     */
-    public function getEngineRecord()
```
Comment (Ostico) [inline lib/Utils/Engines/MMT.php:57]: Nice catch, get engine record should stay in the abstract class.

File: lib/Utils/Validator/EngineValidator.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
+<?php
```
Comment (Ostico) [inline lib/Utils/Validator/EngineValidator.php:1]: We can try to adhere to the Validator interface?

File: lib/Controller/createProjectController.php [hunk 4]
```diff
@@ -128,6 +133,28 @@ public function __construct() {
         $this->only_private            = ( is_null( $this->postInput[ 'get_public_matches' ] ) ? false : !$this->postInput[ 'get_public_matches' ] );
         $this->due_date                = ( empty( $this->postInput[ 'due_date' ] ) ? null : Utils::mysqlTimestamp( $this->postInput[ 'due_date' ] ) );
 
+        // check if MT engine belongs to user
+        if($this->mt_engine > 0 and $this->isLoggedIn()){
+            try {
+                EngineValidator::engineBelongsToUser($this->mt_engine, $this->user->uid);
```
Comment (Ostico) [inline lib/Controller/createProjectController.php:139]: This engineBelongsToUser already instantiate a engine record, there is no need to re-instantiate.

File: lib/Controller/createProjectController.php [hunk 5]
```diff
@@ -128,6 +133,28 @@ public function __construct() {
         $this->only_private            = ( is_null( $this->postInput[ 'get_public_matches' ] ) ? false : !$this->postInput[ 'get_public_matches' ] );
         $this->due_date                = ( empty( $this->postInput[ 'due_date' ] ) ? null : Utils::mysqlTimestamp( $this->postInput[ 'due_date' ] ) );
 
+        // check if MT engine belongs to user
+        if($this->mt_engine > 0 and $this->isLoggedIn()){
```
Comment (Ostico) [inline lib/Controller/createProjectController.php:137]: \> 1

**Review by Ostico — DISMISSED** [review]

File: lib/Controller/createProjectController.php (hunk 4 above)
Comment (mauretto78) [reply lib/Controller/createProjectController.php]: I removed a redundant method. Now this is the only method that fetches the engine.

**Review by riccio82 — APPROVED** [review]

**Review by riccio82 — APPROVED** [review]

**Pull request merged**